featuregates/controller/pkg/feature/feature_controller.go

```diff
@@ -0,0 +1,244 @@
+type FeatureReconciler struct {
+	client.Client
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+}
+
+// +kubebuilder:rbac:groups=core.tanzu.vmware.com,resources=featuregates,verbs=get;list;watch;update;patch
+// +kubebuilder:rbac:groups=core.tanzu.vmware.com,resources=featuregates/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=core.tanzu.vmware.com,resources=features,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=core.tanzu.vmware.com,resources=features/status,verbs=get;update;patch
+
+// Reconcile reconciles the FeatureGate spec by computing activated, deactivated and unavailable features.
+func (r *FeatureReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	ctxCancel, cancel := context.WithTimeout(ctx, contextTimeout)
+	defer cancel()
+
+	log := r.Log.WithValues("feature", req.NamespacedName)
+	log.Info("Starting reconcile")
+
+	feature := &corev1alpha2.Feature{}
+	if err := r.Client.Get(ctxCancel, req.NamespacedName, feature); err != nil {
+		// Reconcile deleted feature
+		if apierrors.IsNotFound(err) {
+			if err := reconcileDeletedFeature(ctx, r.Client, req.NamespacedName.Name); err != nil {
+				return ctrl.Result{}, err
+			}
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	featureGate, found, err := util.GetFeatureGateForFeature(ctx, r.Client, feature.Name)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Reconcile feature that is not found in any FeatureGate spec
+	if !found {
+		if err := reconcileFeatureNotInFeatureGateSpec(ctx, r.Client, feature); err != nil {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+
+	// Reconcile feature that is part of a FeatureGate spec
+	if err := reconcileFeatureInFeatureGateSpec(ctx, r.Client, featureGate, feature); err != nil {
+		return ctrl.Result{}, err
+	}
+	return ctrl.Result{}, nil
+}
+
+// reconcileFeatureInFeatureGateSpec reconciles feature that is present in FeatureGate spec
+func reconcileFeatureInFeatureGateSpec(ctx context.Context, c client.Client, featureGate *corev1alpha2.FeatureGate, feature *corev1alpha2.Feature) error {
+	policy := corev1alpha2.GetPolicyForStabilityLevel(feature.Spec.Stability)
+	featureReference, _ := util.GetFeatureReferenceFromFeatureGate(featureGate, feature.Name)
+	featureResult, activate := applyPolicyToComputeFeatureResultAndActivation(policy, featureReference)
+
+	// Update FeatureGate status
+	featureGate.Status.Results = computeFeatureGateStatusResults(featureGate.Status, featureResult, true)
+	if err := c.Status().Update(ctx, featureGate); err != nil {
+		return fmt.Errorf("could not update %s FeatureGate status :%w", featureGate.Name, err)
+	}
+
+	// Update Feature status
+	feature.Status.Activated = activate
+	if err := c.Update(ctx, feature); err != nil {
+		return fmt.Errorf("could not update %s Feature status :%w", feature.Name, err)
+	}
+	return nil
+}
+
+// reconcileDeletedFeature reconciles feature that has been deleted
```

Comment:
Is this comment referring to the Feature CR being deleted or the FeatureReference being removed from FeatureGate?

Reply:
reconcileDeletedFeature is for reconciling deleted Feature CR, i can expand the comments a little more to make it easy to understand